src/stack_limiter/thunk.rs

```diff
@@ -1,58 +1,103 @@
-#[cfg(not(features = "std"))]
-use alloc::collections::BTreeMap as Map;
-use alloc::vec::Vec;
+use alloc::{collections::BTreeMap as Map, vec::Vec};
 use parity_wasm::{
 	builder,
-	elements::{self, FunctionType, Internal},
+	elements::{self, FunctionType, Instruction, Instructions, Internal},
 };
-#[cfg(features = "std")]
-use std::collections::HashMap as Map;
 
-use super::{resolve_func_type, Context};
+use super::{
+	instrument_call,
+	max_height::{MaxStackHeightCounter, MaxStackHeightCounterContext},
+	resolve_func_type, Context,
+};
 
 struct Thunk {
 	signature: FunctionType,
+	body: Option<Vec<Instruction>>,
 	// Index in function space of this thunk.
 	idx: Option<u32>,
-	callee_stack_cost: u32,
 }
 
-pub fn generate_thunks(
+pub fn generate_thunks<I: IntoIterator<Item = Instruction>>(
 	ctx: &mut Context,
 	module: elements::Module,
-) -> Result<elements::Module, &'static str> {
+	injection_fn: impl Fn(&FunctionType) -> I,
```

Comment:
```suggestion
	injection_fn: impl FnMut(&FunctionType) -> I,
```
May it be more suitable, @SabaunT?

Reply:
It should not mutate captured variables, but should have read access to them. The point of API is customize exit with instructions:
```rust
inject(|signature| [I64Const(0), Call(gr_system_break_idx)])
```